### DISCUSSION TITLE
How to make loader params type safe?

### DISCUSSION THREAD

**Discussion by tykhan** [TanStack/router#1133] [Q&A] [answered]
Assuming I have a router with dynamic parameter 'accountId', which I use in my loader to fetch data.

<img width="631" alt="image" src="https://github.com/TanStack/router/assets/62352586/16d32b4e-f591-435c-aca7-81b6e04c26e6">

The question is, how to make params in my loader type safe automatically?

<img width="631" alt="image" src="https://github.com/TanStack/router/assets/62352586/1888a92e-fc68-4e6b-9e8d-d220ca634023">
<img width="572" alt="image" src="https://github.com/TanStack/router/assets/62352586/dc45ea53-5d55-491b-8cbe-6a3a40503bd6">

I can add a generic to LoaderFnContext, but it's not maintainable since route and loader won't know about changes in each other

<img width="683" alt="image" src="https://github.com/TanStack/router/assets/62352586/f32dcd87-dd51-4fd7-abe8-8626bfa52d7a">

**Comment by tannerlinsley** [comment]
Type your constraints into your overview loader. Then fulfill them when you call it.

**Reply by tykhan** [reply]
Is there any doc referring to that? Not sure If I got what you mean

**Reply by tykhan** [reply]
Not moving a loader logic to a separate function solves this issue, but it's not really convenient

**Reply by tannerlinsley** [reply, marked as answer]
Extracting the loader from the route creates a 3-phase type-dependency chain: **path/search/etc => loader => components (which use loader)**. So while theoretically possible (we tried in prior versions) it explodes the complexity and sometimes even the TS compiler. That said, it's a pretty standard pattern in all TS that if you extract a function from an inference scope, you'll need to type its signature to contain it's required constraints and return value. That's all that's happening here:

```tsx
const Route = createRoute({
  ...
  loader: ({ params }) => fetchPostById(params.postId)
})

function fetchPostById(postId: string): Post {
  ...
}
```
